File: mobile_client/lib/shared/services/socket_service.dart
import 'package:flutter/foundation.dart';
import 'package:socket_io_client/socket_io_client.dart' as io;

class SocketService {
  static SocketService? _instance;

  factory SocketService({String? url}) {
    _instance ??= SocketService._internal(url: url);
    return _instance!;
  }

  SocketService._internal({String? url}) {
    final effectiveUrl = (url == null || url.isEmpty) ? _defaultUrl : url;
    _socket = io.io(
      effectiveUrl,
      io.OptionBuilder()
          .setTransports(['websocket'])
          .disableAutoConnect()
          .build(),
    );

    _setupSocketListeners();
  }

  static const String _defaultUrl = 'http://98.88.78.57:3000';

  late final io.Socket _socket;

  final ValueNotifier<bool> isConnected = ValueNotifier<bool>(false);

  void connect() {
    if (!_socket.connected) {
      _socket.connect();
    }
  }

  void disconnect() {
    _socket.disconnect();
  }

  void on<T>(String event, void Function(T data) callback) {
    _socket.on(event, (data) => callback(data as T));
  }

  void off(String event) {
    _socket.off(event);
  }

  void sendMessage<T>(String event, T message) {
    if (!_socket.connected) {
      _socket.connect();
    }
    if (kDebugMode) {
      debugPrint('socket emit: $event -> $message');
    }
    _socket.emit(event, message);
  }

  String? getSocketId() => _socket.id;

  void dispose() {
    _socket.dispose();
    isConnected.dispose();
  }

  void _setupSocketListeners() {
    _socket.onConnect((_) => isConnected.value = true);
    _socket.onDisconnect((_) => isConnected.value = false);
    _socket.onConnectError((_) => isConnected.value = false);
  }
}

File: client/src/app/services/socket/socket.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { io, Socket } from 'socket.io-client';
import { environment } from 'src/environments/environment';
import { SocketListen } from '@app/enums/socket-enums';

@Injectable({
    providedIn: 'root',
})
export class SocketService {
    private socket: Socket;
    private connectionStatus = new BehaviorSubject<boolean>(false);

    constructor() {
        this.socket = io(environment.socketUrl, {
            transports: [SocketListen.WebSocket],
            autoConnect: false,
        });

        this.setupSocketListeners();
    }
    connect() {
        if (!this.socket.connected) {
            this.socket.connect();
        }
    }
    disconnect() {
        this.socket.disconnect();
    }

    isConnected(): Observable<boolean> {
        return this.connectionStatus.asObservable();
    }
    on<T>(event: string, callback: (data: T) => void) {
        this.socket.on(event, callback);
    }
    off<T>(event: string, callback?: (data: T) => void) {
        this.socket.off(event, callback);
    }
    sendMessage<T>(event: string, message: T): void {
        if (!this.socket.connected) {
            this.socket.connect();
        }
        this.socket.emit(event, message);
    }
    getSocketId(): string {
        return this.socket.id as string;
    }
    private setupSocketListeners() {
        this.socket.on(SocketListen.Connect, () => {
            this.connectionStatus.next(true);
        });

        this.socket.on(SocketListen.Disconnect, () => {
            this.connectionStatus.next(false);
        });

        this.socket.on(SocketListen.ConnectError, () => {
            this.connectionStatus.next(false);
        });
    }
}
